### PR TITLE
GEODE-9186: Add Radish ZPOPMIN command

### PR DESCRIPTION
**Pull request by jdeppe-pivotal**
Thank you for submitting a contribution to Apache Geode.

In order to streamline the review of the contribution we ask you
to ensure the following steps have been taken:

### For all changes:
- [ ] Is there a JIRA ticket associated with this PR? Is it referenced in the commit message?

- [ ] Has your PR been rebased against the latest commit within the target branch (typically `develop`)?

- [ ] Is your initial contribution a single, squashed commit?

- [ ] Does `gradlew build` run cleanly?

- [ ] Have you written or updated unit tests to verify your changes?

- [ ] If adding new dependencies to the code, are these dependencies licensed in a way that is compatible for inclusion under [ASF 2.0](http://www.apache.org/legal/resolved.html#category-a)?

### Note:
Please ensure that once the PR is submitted, check Concourse for build issues and
submit an update to your PR as soon as possible. If you need help, please send an
email to dev@geode.apache.org.

### CODE REVIEW

**Review by dschneider-pivotal — APPROVED** [review]

**Review by upthewaterspout — APPROVED** [review]

File: geode-apis-compatible-with-redis/src/main/java/org/apache/geode/redis/internal/data/NullRedisSortedSet.java [hunk 1]
```diff
@@ -86,6 +86,11 @@ long zcount(SortedSetScoreRangeOptions rangeOptions) {
     return increment;
   }
 
+  @Override
+  List<byte[]> zpopmin(Region<RedisKey, RedisData> region, RedisKey key, int count) {
```
Comment (ringles) [inline geode-apis-compatible-with-redis/src/main/java/org/apache/geode/redis/internal/data/NullRedisSortedSet.java:90]: I think this should go below "zpopmax" alphabetically.

File: geode-apis-compatible-with-redis/src/main/java/org/apache/geode/redis/internal/data/RedisSortedSet.java [hunk 2]
```diff
@@ -401,10 +401,22 @@ long zrevrank(byte[] member) {
     return null;
   }
 
+  List<byte[]> zpopmin(Region<RedisKey, RedisData> region, RedisKey key, int count) {
```
Comment (ringles) [inline geode-apis-compatible-with-redis/src/main/java/org/apache/geode/redis/internal/data/RedisSortedSet.java:404]: Again, we should probably stick to alphabetical order here. Though that means both zpop*'s should go above zscore...

File: geode-apis-compatible-with-redis/src/main/java/org/apache/geode/redis/internal/data/RedisSortedSetCommandsFunctionExecutor.java [hunk 3]
```diff
@@ -103,6 +103,12 @@ public long zrevrank(RedisKey key, byte[] member) {
     return stripedExecute(key, () -> getRedisSortedSet(key, true).zscore(member));
   }
 
+  @Override
+  public List<byte[]> zpopmin(RedisKey key, int count) {
```
Comment (ringles) [inline geode-apis-compatible-with-redis/src/main/java/org/apache/geode/redis/internal/data/RedisSortedSetCommandsFunctionExecutor.java:107]: Another call for alphabetization. :)

File: geode-apis-compatible-with-redis/src/main/java/org/apache/geode/redis/internal/executor/sortedset/RedisSortedSetCommands.java [hunk 4]
```diff
@@ -46,5 +46,7 @@
 
   byte[] zscore(RedisKey key, byte[] member);
 
+  List<byte[]> zpopmin(RedisKey key, int count);
```
Comment (ringles) [inline geode-apis-compatible-with-redis/src/main/java/org/apache/geode/redis/internal/executor/sortedset/RedisSortedSetCommands.java:49]: Gimme an alpha! Gimme a beta! What does that spell? Alphabeta, which is almost 'alpahbet'...

**Review by ringles — CHANGES_REQUESTED** [review]
Just alphabetizing things in a few places.

File: geode-apis-compatible-with-redis/src/integrationTest/java/org/apache/geode/redis/internal/executor/sortedset/AbstractZPopMinIntegrationTest.java [hunk 5]
```diff
@@ -0,0 +1,157 @@
...
+  public void tearDown() {
+    flushAll();
+    jedis.close();
+  }
+
+  @Test
+  public void shouldError_givenWrongNumberOfArguments() {
+    assertThatThrownBy(
+        () -> jedis.sendCommand("key", Protocol.Command.ZPOPMIN, "key", "1", "2"))
+            .hasMessageContaining(RedisConstants.ERROR_SYNTAX);
```
Comment (nonbinaryprogrammer) [inline geode-apis-compatible-with-redis/src/integrationTest/java/org/apache/geode/redis/internal/executor/sortedset/AbstractZPopMinIntegrationTest.java:55]: we have a helper to do this. RedisCommandArgumentsTestHelper

File: geode-apis-compatible-with-redis/src/integrationTest/java/org/apache/geode/redis/internal/executor/sortedset/AbstractZPopMinIntegrationTest.java [hunk 6]
```diff
@@ -0,0 +1,157 @@
...
+    jedis.zadd("key", 1, "player2");
+    jedis.zadd("key", 1, "player3");
+
+    assertThat(jedis.zpopmin("key").getElement()).isEqualTo("player1");
+    assertThat(jedis.zrange("key", 0, 10))
+        .containsExactlyInAnyOrder("player2", "player3");
+  }
+
+  @Test
+  public void shouldReturn_lowestScore() {
```
Comment (nonbinaryprogrammer) [inline geode-apis-compatible-with-redis/src/integrationTest/java/org/apache/geode/redis/internal/executor/sortedset/AbstractZPopMinIntegrationTest.java:91]: I don't think this test name quite captures the intention behind the test. It just needs a little more to it

File: geode-apis-compatible-with-redis/src/integrationTest/java/org/apache/geode/redis/internal/executor/sortedset/AbstractZPopMinIntegrationTest.java [hunk 7]
```diff
@@ -0,0 +1,157 @@
...
+    for (int i = 0; i < count; i++) {
+      jedis.zadd("key", 1D, "player" + shuffles.get(i));
+      tuples.add(new Tuple("player" + i, 1D));
+    }
+
+    assertThat(jedis.zpopmin("key", count)).containsExactlyElementsOf(tuples);
+  }
+
+  @Test
+  public void shouldReturn_countLowestScores() {
```
Comment (nonbinaryprogrammer) [inline geode-apis-compatible-with-redis/src/integrationTest/java/org/apache/geode/redis/internal/executor/sortedset/AbstractZPopMinIntegrationTest.java:130]: this test name also doesn't quite make sense to me

File: geode-apis-compatible-with-redis/src/main/java/org/apache/geode/redis/internal/data/NullRedisSortedSet.java (hunk 1 above)
Comment (jdeppe-pivotal) [reply geode-apis-compatible-with-redis/src/main/java/org/apache/geode/redis/internal/data/NullRedisSortedSet.java]: Done

File: geode-apis-compatible-with-redis/src/main/java/org/apache/geode/redis/internal/data/RedisSortedSet.java (hunk 2 above)
Comment (jdeppe-pivotal) [reply geode-apis-compatible-with-redis/src/main/java/org/apache/geode/redis/internal/data/RedisSortedSet.java]: Done

File: geode-apis-compatible-with-redis/src/main/java/org/apache/geode/redis/internal/data/RedisSortedSetCommandsFunctionExecutor.java (hunk 3 above)
Comment (jdeppe-pivotal) [reply geode-apis-compatible-with-redis/src/main/java/org/apache/geode/redis/internal/data/RedisSortedSetCommandsFunctionExecutor.java]: Done

File: geode-apis-compatible-with-redis/src/main/java/org/apache/geode/redis/internal/executor/sortedset/RedisSortedSetCommands.java (hunk 4 above)
Comment (jdeppe-pivotal) [reply geode-apis-compatible-with-redis/src/main/java/org/apache/geode/redis/internal/executor/sortedset/RedisSortedSetCommands.java]: Done

File: geode-apis-compatible-with-redis/src/integrationTest/java/org/apache/geode/redis/internal/executor/sortedset/AbstractZPopMinIntegrationTest.java (hunk 5 above)
Comment (jdeppe-pivotal) [reply geode-apis-compatible-with-redis/src/integrationTest/java/org/apache/geode/redis/internal/executor/sortedset/AbstractZPopMinIntegrationTest.java:55]: I don't see anything specific in `RedisCommandArgumentsTestHelper` to what this test is asserting.

File: geode-apis-compatible-with-redis/src/integrationTest/java/org/apache/geode/redis/internal/executor/sortedset/AbstractZPopMinIntegrationTest.java (hunk 6 above)
Comment (jdeppe-pivotal) [reply geode-apis-compatible-with-redis/src/integrationTest/java/org/apache/geode/redis/internal/executor/sortedset/AbstractZPopMinIntegrationTest.java]: Done

File: geode-apis-compatible-with-redis/src/integrationTest/java/org/apache/geode/redis/internal/executor/sortedset/AbstractZPopMinIntegrationTest.java (hunk 7 above)
Comment (jdeppe-pivotal) [reply geode-apis-compatible-with-redis/src/integrationTest/java/org/apache/geode/redis/internal/executor/sortedset/AbstractZPopMinIntegrationTest.java]: Hopefully more descriptive now.

File: geode-apis-compatible-with-redis/src/integrationTest/java/org/apache/geode/redis/internal/executor/sortedset/AbstractZPopMinIntegrationTest.java (hunk 5 above)
Comment (DonalEvans) [reply geode-apis-compatible-with-redis/src/integrationTest/java/org/apache/geode/redis/internal/executor/sortedset/AbstractZPopMinIntegrationTest.java:55]: There are two distinct error messages that could be returned for zpopmin (and zpopmax) with the wrong number of arguments. For more than 3 arguments, like in this test, `ERROR_SYNTAX` is returned, but for less than 2 (i.e. missing both the `key` and `count` arguments) then "ERR wrong number of arguments for 'zpopmin' command" is returned. We need tests for both cases, I think.

File: geode-apis-compatible-with-redis/src/test/java/org/apache/geode/redis/internal/SupportedCommandsJUnitTest.java [hunk 8]
```diff
@@ -91,6 +91,7 @@
       "ZINCRBY",
       "ZCARD",
       "ZCOUNT",
+      "ZPOPMIN",
```
Comment (DonalEvans) [inline geode-apis-compatible-with-redis/src/test/java/org/apache/geode/redis/internal/SupportedCommandsJUnitTest.java:94]: Could this be moved to below ZPOPMAX to maintain alphabetical ordering?

File: geode-apis-compatible-with-redis/src/integrationTest/java/org/apache/geode/redis/internal/executor/server/AbstractHitsMissesIntegrationTest.java [hunk 9]
```diff
@@ -252,6 +252,11 @@ public void testZIncrBy() {
     runCommandAndAssertNoStatUpdates("key", (k, m) -> jedis.zincrby(k, 100.0, m));
   }
 
+  @Test
+  public void testZPopMin() {
```
Comment (DonalEvans) [inline geode-apis-compatible-with-redis/src/integrationTest/java/org/apache/geode/redis/internal/executor/server/AbstractHitsMissesIntegrationTest.java:256]: Could this be moved to below `testZPopMax()` to maintain alphabetical ordering?

File: geode-apis-compatible-with-redis/src/test/java/org/apache/geode/redis/internal/data/RedisSortedSetTest.java [hunk 10]
```diff
@@ -473,6 +473,70 @@ public void memberDummyOrderedSetEntryCompareTo_withEqualMemberNamesAndInclusive
...
+    assertThat(result).containsExactly("member1".getBytes(), "1".getBytes());
+
+    ArgumentCaptor<RemsDeltaInfo> argumentCaptor = ArgumentCaptor.forClass(RemsDeltaInfo.class);
+    verify(sortedSet).storeChanges(eq(region), eq(key), argumentCaptor.capture());
+    assertThat(argumentCaptor.getValue().getRemoves()).containsExactly("member1".getBytes());
+    assertThat(rangeSortedSet.getSortedSetSize()).isEqualTo(originalSize - count);
+  }
+
+  @Test
+  public void zpopminRemovesMembersWithHighestScores_whenCountIsGreaterThanOne() {
```
Comment (DonalEvans) [inline geode-apis-compatible-with-redis/src/test/java/org/apache/geode/redis/internal/data/RedisSortedSetTest.java:494]: Should this test name instead be "zpopminRemovesMembersWithLowestScores_whenCountIsGreaterThanOne()"?

File: geode-apis-compatible-with-redis/src/integrationTest/java/org/apache/geode/redis/internal/executor/sortedset/AbstractZPopMinIntegrationTest.java [hunk 11]
```diff
@@ -0,0 +1,157 @@
...
+import org.junit.Test;
+import redis.clients.jedis.HostAndPort;
+import redis.clients.jedis.JedisCluster;
+import redis.clients.jedis.Protocol;
+import redis.clients.jedis.Tuple;
+
+import org.apache.geode.redis.RedisIntegrationTest;
+import org.apache.geode.redis.internal.RedisConstants;
+
+public abstract class AbstractZPopMinIntegrationTest implements RedisIntegrationTest {
```
Comment (DonalEvans) [inline geode-apis-compatible-with-redis/src/integrationTest/java/org/apache/geode/redis/internal/executor/sortedset/AbstractZPopMinIntegrationTest.java:37]: Could we also get a test for the behaviour when the `count` argument is greater than the number of elements in the set?

**Review by DonalEvans — CHANGES_REQUESTED** [review]
A little more alphabetizing, a typo and a couple of missing test cases.

File: geode-apis-compatible-with-redis/src/integrationTest/java/org/apache/geode/redis/internal/executor/sortedset/AbstractZPopMinIntegrationTest.java (hunk 5 above)
Comment (jdeppe-pivotal) [reply geode-apis-compatible-with-redis/src/integrationTest/java/org/apache/geode/redis/internal/executor/sortedset/AbstractZPopMinIntegrationTest.java:61]: Done.

File: geode-apis-compatible-with-redis/src/test/java/org/apache/geode/redis/internal/SupportedCommandsJUnitTest.java (hunk 8 above)
Comment (jdeppe-pivotal) [reply geode-apis-compatible-with-redis/src/test/java/org/apache/geode/redis/internal/SupportedCommandsJUnitTest.java]: Done

File: geode-apis-compatible-with-redis/src/integrationTest/java/org/apache/geode/redis/internal/executor/server/AbstractHitsMissesIntegrationTest.java (hunk 9 above)
Comment (jdeppe-pivotal) [reply geode-apis-compatible-with-redis/src/integrationTest/java/org/apache/geode/redis/internal/executor/server/AbstractHitsMissesIntegrationTest.java]: Done

File: geode-apis-compatible-with-redis/src/test/java/org/apache/geode/redis/internal/data/RedisSortedSetTest.java (hunk 10 above)
Comment (jdeppe-pivotal) [reply geode-apis-compatible-with-redis/src/test/java/org/apache/geode/redis/internal/data/RedisSortedSetTest.java]: Done.

**Review by DonalEvans — APPROVED** [review]

File: geode-apis-compatible-with-redis/src/integrationTest/java/org/apache/geode/redis/internal/executor/sortedset/AbstractZPopMinIntegrationTest.java (hunk 11 above)
Comment (jdeppe-pivotal) [reply geode-apis-compatible-with-redis/src/integrationTest/java/org/apache/geode/redis/internal/executor/sortedset/AbstractZPopMinIntegrationTest.java:38]: Done.

**Review by nonbinaryprogrammer — APPROVED** [review]

**Review by ringles — APPROVED** [review]

**Pull request merged**